drivers/SmartThings/zigbee-fan/src/kichler-fan-light/init.lua

```diff
@@ -0,0 +1,187 @@
+-- Copyright 2024 SmartThings
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+local clusters = require "st.zigbee.zcl.clusters"
+local capabilities = require "st.capabilities"
+local FanControl = clusters.FanControl
+local Level = clusters.Level
+local OnOff = clusters.OnOff
+
+local FINGERPRINTS = {
+  { mfr = "KICHLER", model = "KICHLER-FANLIGHT-Z-301" },
+}
+
+local function can_handle_itm_fanlight(opts, driver, device)
+  for _, fingerprint in ipairs(FINGERPRINTS) do
+    if device:get_manufacturer() == fingerprint.mfr and device:get_model() == fingerprint.model then
+      return true
+    end
+  end
+  return false
+end
+
+-- CAPABILITY HANDLERS
+local function on_handler(driver, device, command)
+  local last_level = device:get_field('LAST_DIM_LEVEL') or 100
+  local level = math.floor((last_level/100.0) * 254 )
+  if command.component == 'light' then
+    device:send_to_component('light', Level.server.commands.MoveToLevelWithOnOff(device, level, command.args.rate or 0xFFFF))
+    device:set_field('LAST_DIM_LEVEL', last_level, {persist = true})
+  else
+    local speed = device:get_field('LAST_FAN_SPD') or 1
+    device:send(FanControl.attributes.FanMode:write(device, speed))
+  end
+  device:send(FanControl.attributes.FanMode:read(device))
+end
+
+local function off_handler(driver, device, command)
+  if command.component == 'light' then
+    local last_level = device:get_field('LAST_DIM_LEVEL') or 100
+    --local level = math.floor((last_level/100.0) * 254 )
+    device:send_to_component('light', Level.server.commands.MoveToLevelWithOnOff(device, 0, command.args.rate or 0xFFFF))
+    device:set_field('LAST_DIM_LEVEL', last_level, {persist = true})
+  else
+    device:send(FanControl.attributes.FanMode:write(device, FanControl.attributes.FanMode.OFF))
+  end
+  device:send(FanControl.attributes.FanMode:read(device))
+end
+
+local function switch_level_handler(driver, device, command)
+  local trim_level = tonumber(device.preferences.trim) or 10
+  if command.args.level <= trim_level and command.args.level >= 1 then
+    local level = math.floor((trim_level/100.0) * 254 )
+    device:emit_component_event(device.profile.components.light, capabilities.switchLevel.level(command.args.level))
+    device:send_to_component('light', Level.server.commands.MoveToLevelWithOnOff(device, level, command.args.rate or 0xFFFF))
+    device:emit_component_event(device.profile.components.light, capabilities.switchLevel.level(trim_level))
```

Comment:
What's the intent behind these two `level` events from the same component here?

Reply:
This is to use the Trim function. This trim is used to limit the minimum value of dimming.

Reply:
I understand the point of the trim function, but it seems like you'd want to only emit the trimmed value, not both the untrimmed and trimmed value.

Reply:
I would like to refresh the value below the trim setting value so that it is not reflected in the UI. If you set it to trim 25% and control it to 10%, it will be 25%.

Reply:
So why send both the 10% event and the 25% event? Wouldn't you want to just send the 25% event?

Reply:
![24캡처](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/assets/136560804/fd24c708-5203-4771-9320-32e5c1b1daff)

Set Trim to 25%.
If the user has changed to 10% in the UI, the intention is to express it as 25%.
However, if you adjust 25% and then adjust it to less than that (e.g., 20%), it is to inform you of the event that it has been changed to go back to 25%